src/core/src/model.cpp

```diff
@@ -215,6 +215,15 @@ void ov::Model::validate_nodes_and_infer_types() const {
     std::stringstream unregistered_parameters;
     std::stringstream unregistered_variables;
     std::unordered_set<const ov::descriptor::Tensor*> tensors;
+
+    // Next containers are used to fix collisions in autogenerated names
+    // Collect all unique friendly names
```

Comment:
I would separate this logic from function to dedicated transformation (like in ModelOptimizer) because all this checks are not for free and potentially will increase load time. In average we have 5 function validations per plugin pipeline. Also we can't guaranty that during transformation pipeline we will have function validations even if transformations change graph. So we have manually call this method in the end. Also ngraph Function shouldn't guaranty unique names for nodes as this is just plugin wish. So considering all above I would separate this logic to dedicated transformation and put it to the end of transformation pipeline.